### PR TITLE
Tokenized Input Feature

### PR DESCRIPTION
**Pull request by JiahuiKChen**
Tokenizer class and corresponding test suite are finished. All JUnit tests for Tokenizer class pass. How the command line commands are parsed was changed from using a Scanner to using the Tokenizer and a List of Strings (these changes are in the Panel and Chat classes). Some uncertainty (marked with TODO comments) with how some code should be changed over, the command line parsing also needs testing, but I'm not sure how to do that.

### CODE REVIEW

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -47,9 +46,22 @@ public Chat(Context context) {
...
 
-    final Scanner tokens = new Scanner(line.trim());
+    final List<String> args = new ArrayList<>();
 
-    final String command = tokens.hasNext() ? tokens.next() : "";
+    final Tokenizer tokenizer = new Tokenizer(line);
+
+    //tokenizing the input, and adding each token to the ArrayList
+    try {
+      for (String token = tokenizer.next(); token != null; token = tokenizer.next()) {
```
Comment (kevmo314) [inline src/codeu/chat/client/commandline/Chat.java]: It seems like it would be useful if Tokenizer implemented `Iterator<String>`, that way you could just do `tokenizer.forEachRemaining(args::add);`

File: src/codeu/chat/client/commandline/Chat.java [hunk 2]
```diff
@@ -136,8 +148,11 @@ public void invoke(Scanner args) {
     //
     panel.register("u-add", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        //TODO: possible place of error, not sure how to test username creation
+        // what's in the List being passed in? Is it already tokenized? The scanner code took the entire line?
+        // final String name = args.hasNext() ? args.nextLine().trim() : ""; (old/given code, uses a parameter Scanner args)
+        final String name = args.get(0) != null ? args.get(0) : ""; //this works if the args list passed in contains tokens
```
Comment (kevmo314) [inline src/codeu/chat/client/commandline/Chat.java]: `args.size() > 0` seems like a better check than `args.get(0) != null`. Technically, `get()` isn't guaranteed to return null: https://docs.oracle.com/javase/8/docs/api/java/util/List.html#get-int-

Same for the remaining instances.

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (ditadewindita) [inline src/codeu/chat/client/commandline/Chat.java]: Wont calling `tokenizer.next()` twice result in skipped over inputs?

Try using a while loop:
```
String token;
while((token = tokenizer.next()) != null)
     args.add(token);
```

File: src/codeu/chat/util/Tokenizer.java [hunk 3]
```diff
@@ -1,12 +1,134 @@
...
+     * characters in the source String for every method call.
+     *
+     * @return  String      Next group of characters without white space in the source String.
+     * @throws IOException  May throw an IO Exception through calling the read() method.
+     */
     public String next() throws IOException {
+        //ignores all whitespace at beginning of the source String
+        while (remaining() > 0 && Character.isWhitespace(peek())){
+            read();
+        }
```
Comment (ditadewindita) [inline src/codeu/chat/util/Tokenizer.java]: Easier to use `source.trim()` to skip over leading and trailing whitespace

File: src/codeu/chat/client/commandline/Chat.java [hunk 4]
```diff
@@ -136,8 +148,11 @@ public void invoke(Scanner args) {
     //
     panel.register("u-add", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        //TODO: possible place of error, not sure how to test username creation
+        // what's in the List being passed in? Is it already tokenized? The scanner code took the entire line?
```
Comment (kevmo314) [inline src/codeu/chat/client/commandline/Chat.java]: To make it clearer what `List<string>` contains, could you instead just pass a `Tokenizer` to all these methods? It seems like it just calls with whatever is passed to `.handleCommand()` up on line 81. I'm not sure if it would work though, either is fine.

File: src/codeu/chat/util/Tokenizer.java [hunk 5]
```diff
@@ -1,12 +1,134 @@
...
 
-    public Tokenizer(String source) {
+    //stores the final "token" form of the input
+    private StringBuilder token;
+
+    //the input being processed/tokenized
+    private String source;
 
+    //index where current character being read is located
+    private int at;
```
Comment (ditadewindita) [inline src/codeu/chat/util/Tokenizer.java]: Better variable name? Maybe `currentIndex` ?

File: src/codeu/chat/util/Tokenizer.java [hunk 6]
```diff
@@ -1,12 +1,134 @@
...
+     * @return  String      Next group of characters without white space in the source String.
+     * @throws IOException  May throw an IO Exception through calling the read() method.
+     */
     public String next() throws IOException {
+        //ignores all whitespace at beginning of the source String
+        while (remaining() > 0 && Character.isWhitespace(peek())){
+            read();
+        }
+        //returns null if there are no more characters in the String
+        if (remaining() <= 0 ){
```
Comment (ditadewindita) [inline src/codeu/chat/util/Tokenizer.java]: If using `.trim()`, you could just check if the length of the source string is `0` after trimming the whitespace

File: src/codeu/chat/util/Tokenizer.java [hunk 7]
```diff
@@ -1,12 +1,134 @@
...
+    }
+
+    /**
+     * Returns the next group of characters in the source String
+     * with no surrounding quotes for every method call.
+     *
+     * @return  String      Next group of characters without white space in the source String.
+     * @throws IOException  May throw an IO Exception through calling the read() method.
+     */
+    private String readWithNoQuotes() throws IOException {
```
Comment (kevmo314) [inline src/codeu/chat/util/Tokenizer.java]: You could combine `readWithNoQuotes` and `readWithQuotes` to a function that looks like

```
private String readUntil(Function<Character, Boolean> terminationCondition) {
   ... append until terminationCondition(peek()) is true ...
}
```

then your code in `next()` becomes

```
else if (peek() == '"') {
  read();
  readUntil(c -> c == '"');
  read();
} else {
  readUntil(c -> !Character.isWhitespace(c));
}
```

This should make your code a bit shorter too.

File: src/codeu/chat/util/Tokenizer.java [hunk 8]
```diff
@@ -1,12 +1,134 @@
...
+        //ensures first character is quotation, also allows next non-quote character to be read next
+        if (read() != '"'){
+            throw new IOException("String is not surrounded by quotes!");
+        }
+        //appends characters until a closing quotations mark is reached
+        while (remaining() > 0 && peek() != '"'){
+            token.append(read());
+        }
+        //reads the closing quotation mark
+        read();
```
Comment (ditadewindita) [inline src/codeu/chat/util/Tokenizer.java]: Input string could have a beginning quote with no end quote.
This method can also be done without a StringBuilder.

Try:
```
if(source.charAt(0) != '\"' && source.charAt(source.length() - 1) != '\"')
     throw new IOException("String is not surrounded by quotes!");
return source;
```

File: test/codeu/chat/util/TokenizerTest.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        assertEquals("4", noQuotesSimple.next());
+        assertEquals(null, noQuotesSimple.next());
+
+        final Tokenizer noQuotesEmpty = new Tokenizer("");
+        assertEquals(null, noQuotesEmpty.next());
+    }
+
+    @Test
+    public void testWithQuotes() throws IOException{
+        final Tokenizer withQuotesSimple = new Tokenizer("\"hello world\" \"how are you\"");
```
Comment (kevmo314) [inline test/codeu/chat/util/TokenizerTest.java:26]: What happens if you have a pair of unmatched quotes, so something like `"\"Some string starting with a quote"`?

Does the technical activities doc mention how robust this class is supposed to be? If it doesn't say anything about catching all edge cases then I wouldn't worry about it.

**Review by kevmo314 — CHANGES_REQUESTED** [review]
A couple comments, overall looks good!

File: test/codeu/chat/util/TokenizerTest.java (hunk 9 above)
Comment (JiahuiKChen) [reply test/codeu/chat/util/TokenizerTest.java:26]: I thought of edge cases like that too, but the technical activities doc didn't mention any edge cases and also the example test cases didn't cover edge cases so I figured it was okay to leave it as is.

File: src/codeu/chat/client/commandline/Chat.java (hunk 4 above)
Comment (JiahuiKChen) [reply src/codeu/chat/client/commandline/Chat.java]: If the `List<String>` passed in contains the output of `.handleCommand()` then it should just be a list of tokens from a String that's already been tokenized, so I was hoping that's the case. The issue with passing a `Tokenizer` into each method is that there needs to be a String passed into `Tokenizer` and the input `String line` is passed into `.handleCommand()` but not saved as a variable for other methods to use.

File: src/codeu/chat/client/commandline/Chat.java (hunk 2 above)
Comment (JiahuiKChen) [reply src/codeu/chat/client/commandline/Chat.java]: Thanks, I will change all the instances of this check.

File: src/codeu/chat/util/Tokenizer.java (hunk 7 above)
Comment (JiahuiKChen) [reply src/codeu/chat/util/Tokenizer.java]: Great! This will also help me become more familiar with the use of Lambda Expressions.

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (JiahuiKChen) [reply src/codeu/chat/client/commandline/Chat.java]: Alright, thanks.

File: src/codeu/chat/client/commandline/Chat.java (hunk 4 above)
Comment (kevmo314) [reply src/codeu/chat/client/commandline/Chat.java]: Hmm, I don't think I understand. Are you not able to change line 81 to `if (panels.peek().handleCommand(command, tokenizer)) {` and remove the entire loop and `args` construction above it?

**Review by kevmo314 — CHANGES_REQUESTED** [review]

**Review by ditadewindita — CHANGES_REQUESTED** [review]
Looks good! Only concern is some of the IDE-specific config files are committed; they must not have been ignored by the `.gitignore` file.

**Review by ditadewindita — APPROVED** [review]
